api/envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.proto

```diff
@@ -21,19 +22,65 @@ message TcpProxy {
-  //  Otherwise, a :ref:`deprecated_v1
-  //  <envoy_api_field_config.filter.network.tcp_proxy.v2.TcpProxy.deprecated_v1>` configuration is
-  //  required to use more complex routing in the interim.
-  //
-  string cluster = 2;
+  // [#not-implemented-hide:] Weighted clusters allow for specification of
+  // multiple upstream clusters along with weights that indicate the
+  // percentage of traffic to be forwarded to each cluster. TCP Proxy
+  // selects an upstream cluster based on the weights.
+  message WeightedCluster {
+    message ClusterWeight {
+      // Name of the upstream cluster. The cluster must exist in the
+      // :ref:`cluster manager configuration <config_cluster_manager>`.
+      string name = 1 [(validate.rules).string.min_bytes = 1];
+      
+      // An integer between 0 and :ref:`total_weight
+      // <envoy_api_field_route.WeightedCluster.total_weight>`. The choice of an upstream
+      // cluster is determined by its weight. The sum of weights across all entries in the
+      // clusters array must add up to the total_weight, which defaults to 100.
+      google.protobuf.UInt32Value weight = 2;
+
+      // [#not-implemented-hide:] Optional endpoint metadata match
+      // criteria. Only endpoints in the upstream cluster with metadata
+      // matching that set in metadata_match will be considered. The filter
+      // name should be specified as *envoy.lb*.
+      core.Metadata metadata_match = 3;
+
+      // [#not-implemented-hide:] The per_filter_config field can be used
+      // to provide weighted cluster-specific configurations for
+      // filters. The key should match the filter name, such as
+      // *envoy.buffer* for the HTTP buffer filter. Use of this field is
+      // filter specific; see the :ref:`HTTP filter documentation
+      // <config_http_filters>` for if and how it is utilized.
+      map<string, google.protobuf.Struct> per_filter_config = 4;
+    }
+
+    // Specifies one or more upstream clusters associated with the TCP proxy.
+    repeated ClusterWeight clusters = 1 [(validate.rules).repeated .min_items = 1];
+
+    // Specifies the total weight across all clusters. The sum of all cluster
+    // weights must equal this value, which must be greater than 0. Defaults
+    // to 100.
+    google.protobuf.UInt32Value total_weight = 3 [(validate.rules).uint32.gte = 1];
+  }
+  
+  oneof cluster_specifier {
+    option (validate.required) = true;
+
+    // The upstream cluster to connect to.
+    string cluster = 2 [(validate.rules).string.min_bytes = 1];
+
+    // [#not-implemented-hide:] Envoy will determine the cluster to route
+    // to based on the SNI presented by the client during the TLS
+    // handshake. If the SNI is empty (as is the case for plaintext TCP
+    // connections) or the referenced cluster does not exist, Envoy will
+    // terminate the connection.
+    google.protobuf.BoolValue use_sni_as_cluster = 10;
```

Comment:
I think we should do this with a more generic approach. We had a discussion last week about allowing per-connection metadata, which filters could set and consume. I think we should have tcp_proxy consume something that is a cluster override value, and write a separate filter that sets this value based on SNI. Then we can support arbitrary routing of this type without putting it all hard-coded into tcp_proxy.

Reply:
Also, in https://github.com/envoyproxy/envoy/issues/4357, it was suggested to disallow cluster names with '.' in them, which would prevent a cluster from being named with the SNI value.

Reply:
The per connection metadata - is this same as the request info stuff that went into HTTP conn. manager?

Reply:
Discussion was here: https://github.com/envoyproxy/envoy/pull/4331

Reply:
Ah yes. But setting cluster from another filter means the tcp proxy can no longer be the terminating filter in the network stack. We need to have http style router as the final filter. And it’s unclear how it would work with things like redis.

Reply:
No, I think tcp_proxy would still be the terminating filter. Something setting metadata would come before it in the filter chain.

For SNI, the thing before it could set the metadata in `onNewConnection()`. For something inspecting the network traffic, it may require changes to tcp_proxy to not connect to upstream until it gets the message for what to connect to.

Reply:
I don’t understand. How will the filter set the cluster that tcp proxy should use to route to upstream? The metadata stuff works for logging purposes. Not for changing the route that tcp proxy is supposed to take.

Reply:
Yes I know about that request info stuff. 

But I would much rather opt for more explicit semantics than an implicit “if not set, use the cluster in tcp proxy”. It makes it hard to debug and the configuration at first sight doesn’t tell you what mode Envoy is using. 

Put another way, we might as well make the cluster field optional so that whatever filter I add earlier could set the cluster, allowing me to leave the cluster field in tcp proxy to be empty. Even then, it feels a bit awkward.

Reply:
Yeah, I agree it should be allowed to leave the cluster empty.

What I'm trying to avoid is everyone putting their own special routing rules hard-coded into tcp_proxy. Especially when it adds opaque requirements on other parts of the configuration (like cluster names must match SNI values).

Another approach we could think about is adding a hook for a lua script to run and select the cluster to use.

Reply:
I am okay with the separate filter. Just don't want to implicitly pull and override a user specified config in tcp proxy.